extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/DurableStorageCleaner.java

```diff
@@ -88,13 +88,22 @@ public void run() throws Exception
                                            .map(TaskRunnerWorkItem::getTaskId)
                                            .map(DurableStorageUtils::getControllerDirectory)
                                            .collect(Collectors.toSet());
+    Set<String> knownTaskIds = taskRunner.getKnownTasks()
+                                         .stream()
+                                         .map(TaskRunnerWorkItem::getTaskId)
+                                         .map(DurableStorageUtils::getControllerDirectory)
+                                         .collect(Collectors.toSet());
 
     Set<String> filesToRemove = new HashSet<>();
     while (allFiles.hasNext()) {
       String currentFile = allFiles.next();
-      String taskIdFromPathOrEmpty = DurableStorageUtils.getControllerTaskIdWithPrefixFromPath(currentFile);
-      if (taskIdFromPathOrEmpty != null && !taskIdFromPathOrEmpty.isEmpty()) {
-        if (runningTaskIds.contains(taskIdFromPathOrEmpty)) {
+      String nextDirName = DurableStorageUtils.getNextDirNameWithPrefixFromPath(currentFile);
+      if (nextDirName != null && !nextDirName.isEmpty()) {
+        if (runningTaskIds.contains(nextDirName)) {
+          // do nothing
+        } else if (DurableStorageUtils.QUERY_RESULTS_DIR.equals(nextDirName)
```

Comment:
Can you please confirm if this logic works? Let's say __query_results has a known id and an unknown id, then we would execute `files.remove(currentFile)` when we iterate over the unknown id.
However the `currentFile` would be the whole results directory or am I wrong?

Reply:
I did not understand what you meant here, current file would be something inside "unknown id", so it would add that file to the list to delete. (We only delete individual files in the cleaner, the storage connector does not allow us to delete directories with the function called)